Custom dialog boxes!

frankitox16 opened the pull request:
This is an approach to custom dialog boxes and a way to be called by LUA functions (that will be later added). It uses an XML File to define elements that will be part of the content of the dialog box, yet i've got to decide which kind of elements should a custom dialog box have (like a "FurnitureSelect" element or buttons, or things like that).

I also added a dictionary in the DialogBoxManager so that lua functions can call dialog boxes by name (not implemented yet)

Right now if you test it, this just adds the dialog box to de canvas, and updates it's title, however the whole TODO would be:

- [x] Add lua actions for custom dialog box.
- [x] Add more content for dialog boxes.
- [x] Allow LUA to show custom dialog boxes.
- [x] Add callback for custom dialog boxes.
- [x] Implement DialogBoxResult in custom dialog boxes.

To see the example that I've added, ~~you have to pause the game, and under Canvas->DialogBoxes you should see a "Testing" dialog box. Enable it~~ Right click a Stockpile and you'll see a "Set Filter" button. Press it and the dialog box that was loaded from the StreamingAssets/UI/DialogBoxes folder will appear. If you want to play a little bit with it, I'm making a [Wiki](https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-does-it-work%3F#adding-external-moddable-ui-dialog-box-frankitox16) page to show how it works and how to add your custom ones.



Bonus! A little screenshot of the example:
![image](https://cloud.githubusercontent.com/assets/15928972/19195181/d634293a-8c86-11e6-84ac-dd819b15f06b.png)

dusho commented:
any chance you can use De/serialization instead of XmlReader? We won't do much about code we have now, but if any future code can use deserialization, we can port it easily to other data language if needed..
Also I wouldn't use DB instead of DialogBox.. (DialogBox isn't that long word) may be bit confusing for all the people knowing and using DB for database stuff..

frankitox16 commented:
@dusho hmm I don't have much experience with de/serialization. Is it something related with that XmlSerializable tag some documents have? About the DB stuff, yes I'll change it to DialogBox when I get home.

dusho commented:
it is actually much easier than all XmlReader magic, easy example here:
https://support.microsoft.com/en-us/kb/815813
with way to customize the output:
http://stackoverflow.com/questions/5937630/serialize-class-to-xml

dusho commented on Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs:
```diff
@@ -0,0 +1,19 @@
+
+public class DialogComponent
+{
+    [XmlAttribute("Type")]
```
according to guidelines here, we have elements starting with capitals and attributes are camelCase, so:
`[XmlAttribute("type")]`

dusho commented on Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs:
```diff
@@ -0,0 +1,19 @@
+    [XmlElement]
+    public Vector3 position;
+
+    [XmlElement]
```
[XmlAttribute("Data")]

dusho replied:
ah ye, sorry.. copy-paste failed me, just wanted to put there capitals
[XmlElement("Data")]

dusho commented on Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs:
```diff
@@ -0,0 +1,19 @@
+    [XmlAttribute("Type")]
+    public string ObjectType;
+
+    [XmlElement]
```
[XmlAttribute("Position")]
shame here that Unity doesn't serialize this with x,y,z as attributes. Though use can create wrapper class, adjust the format then have it return Unitys Vector3..

bjubes commented on Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs:
```diff
@@ -0,0 +1,162 @@
+        CloseDialog();
+    }
+
+    public void OkButtonClicked()
```
you call the OnClosed trigger in each __ButtonClicked method, then call closeDialog in those methods, which calls the same trigger again. Remove those OnClosed trigger calls in the specific buttons but leave the call in the CloseDialog() Method

frankitox16 replied:
Whoops... begginer mistake

bjubes commented on Assets/Scripts/UI/DialogBox/PromptOrInfo/DialogBoxResult.cs:
```diff
@@ -10,5 +10,6 @@ public enum DialogBoxResult
     No,
-    Cancel
+    Cancel,
+    OK
```
OK -> Okay

frankitox16 commented:
Okay, almost finished. Would like to add some more "components" like images or Inventory selection, but I'm tempted to leave it as it is, and allow people to add their stuff, as this isn't exactly a v0.1 feature.

TomMalbran commented:
I don't like the use of Lua in the names. Dialogs could be just Xml or later we could decide not to use Lua. The name shouldn't mention the implementation.

bjubes commented on Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs:
```diff
@@ -0,0 +1,159 @@
+    
+    public override void ShowDialog()
+    {
+        base.ShowDialog();
```
Suggestion: I would like a way to reference show dialog in addition to  close dialog from lua. that seems like a very likely thing to need, say if you wanted to have a sound play when a dialog opens.

frankitox16 commented:
@bjubes I've added the ShowDialogByName method and made it available for it's use on LUA

bjubes commented:
Now that this is no longer WIP, can you specify how to test this? I don't see the test dialog you metioned in your OP so is there a way to open that dialog? or did you implement an existing dialog using the new lua way?

frankitox16 commented:
@bjubes Whoops! Should've added it in my description when i edited it. Anyway, if you pause the game and look in the dialog boxes container in the canvas, you'd see a "Testing" dialog box, which was loaded by the script. I'm in the process of adding how this works in the wiki, but knowing this you should be able to test it. 
The XML and lua files are in the StreamingAssets folder, under UI/DialogBoxes

TomMalbran commented on Assets/Scripts/Models/DialogBoxInformation/DialogBoxLuaInformation.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Xml.Serialization;
+using UnityEngine;
+
+public class DialogBoxLuaInformation
```
Don't name the classes with Lua. The implementation shouldn't be part of the name. Same goes for all the other names that have Lua, like the prefab, and the DialogBoxLua.

bjubes replied:
Maybe use the word mod instead, so ModDialogBox

frankitox16 replied:
I think I've got it. Please tell me if you find any more issues.

bjubes commented:
a few things i noticed:
- the buttons on the dialog are different than the other dialogs (im basing that off of the dialogs dynamically created in code for confirming quit, creating a new world, etc...)
- when i use this code:
``` 
<Title>Testing</Title>
  <Content>
  	<Text>this is sample text</Text>
  </Content>
  <Buttons>Yes</Buttons>
... and so on
```
in testDB.xml no text shows up.
-  there is no way (or i am missing it ) to interface with dialogBoxPromptOrInfo using lua to make generic dialog boxes. this would be helpful when a mod author doesn't want to define multiple xml files just to change the text inside a simple popup. this may be as easy as adding [MoonSharpUserData] to to the top of the dialogBoxPromptOrInfo class

frankitox16 commented:
@bjubes 
Ok, I'll be trying to add the interaction with dialog prompts. As for the dialog buttons, could you please add an image showing the difference? I don't seem to get it.

Text doesnt show up because the `<content>` tag requires a "type" argument to specify what kind of content it is ([reference](https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-does-it-work%3F#adding-external-moddable-ui-dialog-box-frankitox16)), so in order to add text, you'd add: 
```
...
<Title>Testing</Title>
<Content name="customName" type="Text">
<Position>
<x></x>
<y></y>
<z></z>
</Position>
<Data>this is sample text</Data>
</Content>
...
```
The bad thing you made me notice is that `<Content>` is a terrible name for that, as it's not descriptive enough.

bjubes commented:
![capture](https://cloud.githubusercontent.com/assets/7608114/19240536/db289526-8ed8-11e6-848e-f317f36cb077.PNG)
i just noticed the button size is inconsitent. however, i don't know if there is a actual "standard' size in the project or if all the buttons are varying sizes

Tranberry commented:
some style guidelines would be good imo.

frankitox16 commented:
@bjubes I've added the interface with not only "prompt or info" dialog box, but also with the others (however not all of them implement it yet) and corrected the buttons to match prompt's button height.

@Tranberry would you give me an example of the style guidelines that could be added? So I could get a better idea of what you mean...

Also, special thanks to @koosemose and @TomMalbran for the help with the Lua interface, you helped me a lot!

bjubes commented:
Changes look good from the diff, but now there are merge conflicts. Once you resolve them change the title and ill test this out!

bjubes commented:
Sorry, I just merged in a PR before checking this one out so there are conflicts again. Ill make sure to test this one next after the conflicts are fixed

Ohda commented on Assets/Scripts/UI/DialogBox/DialogBoxManager.cs:
```diff
@@ -39,42 +39,66 @@ public void Awake()
+            AddMainMenuItems();
+        }else
+        {
+            tempGoObj = CreateDialogGO("DB_SaveFile", "Save File");
```
There is a problem with your merge conflict here. You don't load the good GameObject on the good scenes.

frankitox16 replied:
You mean the right* game objects on the right scenes, am I right? :sweat_smile: I found the difference with the actual code, the conflicts weren't exactly clear. I'll fix them right away

Ohda replied:
French guys are bad with english. Sorry mate :)

koosemose commented:
Looks mostly good. One minor nitpick, in the test dialog when you choose yes and it brings up the PromptOrInfo dialog, it continues closing (which also leaves the PromptOrInfo dialog as nonmodal), which is probably the right thing in most scenarios, however in the specific case of an "Are You Sure", I think the proper behavior would be to delay closing until a response is received from PromptOrInfo, and either continue closing on Yes or just not close on No. 

Is the system currently capable of having behavior like this and it just isn't wired up for the test case or is it not currently possible to do that. If it isn't that would be a nice addition if it isn't too much extra code (if it is a lot it could be done in a follow-up PR).

frankitox16 commented:
@koosemose Now that I've added listeners to the buttons, we're able to keep the DialogBox before closing the prompt.

koosemose commented:
So that I know what to expect for testing, is the test made to do this, or is it just that the capability is generally there?

frankitox16 commented:
The "filter" test is just a way to prove the way it works, and how you'd modify it. However, that filter dialog may become the dialog that will set stockpile filters, when I figure out a proper way to select inventory/set priorities. As for the listeners, are for every modded dialog box, not for all of them (eg: the settings menu one)

koosemose commented:
I mean the behavior of the dialog staying open during the "Are you sure" Dialog

frankitox16 commented:
That's why I changed the listeners. Before they would trigger an event AND close the dialog. Now they just trigger an event, so the closing of the dialog should be handled inside lua. Long story short: the prompt dialog will be displayed while the dialog is still shown, you can get a result from the prompt and act in consecuence, and in order to close the dialog, you must do it from lua

koosemose commented:
Alright, just wanted to make sure my expectations of how things will work is correct, to minimize time wasted if I misunderstand how it's intended to perform and think it's not performing as it should. :)

koosemose commented:
Looks good, but you've got some stylecop issues, once those are fixed we should be ready for merging (assuming of course nothing breaks unexpectedly)

koosemose commented:
Conflicted

koosemose commented:
Whoops wrong PR

frankitox16 commented:
StyleCop fixed. All clean

koosemose commented:
You seem to have missed ModDialogBoxInformation as well.

frankitox16 commented:
done :sweat_smile:

koosemose commented:
Looks good, stylecop clean, no new bugs rearing their nasty heads. Merged!
